Simplify getting model stats and add more stats

cdcapano opened the pull request:
Work in progress.

Commit:
use ModelStats in Models

Commit:
provide default stats that are returned

Commit:
update analytic models

Commit:
update gaussian noise models

Commit:
move BaseDataModel to it's own file

Commit:
remove model stats to arrays function (probably not needed)

Commit:
some cleanup

Commit:
add CallModel wrapper class

Commit:
simplify from_config call

Commit:
bug fixes

Commit:
fix some bugs

Commit:
update docs some

pep8speaks commented:
Hello @cdcapano! Thanks for updating the PR.



























Cheers ! There are no PEP8 issues in this Pull Request. :beers: 

##### Comment last updated on July 17, 2018 at 15:40 Hours UTC

Commit:
whitespace

Commit:
add functions to retrieve optimal SNRsq, cplx_loglr, and lognl in each detector

Commit:
promote attributes of the model to the call_model's namespace

Commit:
update sampler.from_cli to use CallModel

Commit:
update kombine

Commit:
update sampler base

Commit:
allow callfunc, return_all_stats to be overridden with kwargs

Commit:
fix typo

Commit:
update emcee_pt's global call functions

Commit:
update models unit tests

Commit:
update sampler unit tests

Commit:
fix whitespace

Commit:
fix whitespace

Commit:
add common function for trying to get stats from current_stats

Commit:
fix whitespace

cdcapano commented:
This simplifies the models by moving everything about default call functions to a separate wrapper class called `CallModel`. This also allows supplementary statistics to be returned by the model. Now, e.g., `GaussianNoise` will store the optimal SNR squared in each detector, along with the complex log likelihood ratio, when `loglr` is called. These are then passed as apart of the `blob` data.

To make this work, the models can no longer be called as functions. Instead, you have to first run `.update` with a set of parameter values. You then get various statistics, e.g., `loglikelihood`, with `model.loglikelihood`, which is a property. The first time this is called after a `.update` a function associated with the property is called to calculate the statistic. For example, in the case of `loglikelihood` the associated function is `_loglikelihood`. The function will store its statistic (plus any other supplementary statistics; see `GaussianNoise._logplr` for example) to an instance of `ModelStats`. `ModelStats` stores all statistics that have been calculated for the current point. A new instance is created on each `.update`. All of these statistics can be retrieved as a dict using the `current_stats` attribute, or as a tuple using the `get_current_stats` method. The stats that are returned by default are specified by the model's `default_stats` attribute. Any stats that were not calculated (e.g., if the parameters are non-physical, preventing calculation of the log likelihood) are returned as nans.

The `CallModel` class can be used to wrap a model and call it like a function, so the models can be used with samplers like  `emcee` and `kombine`. Future samplers can then either use the `CallModel` wrapper, or call the model directly to get whatever statistic they need.

As an example, here is how you now use `TestNormal`:
```
    >>> m = TestNormal(['x', 'y'])
    >>> m.update(x=-0.2, y=0.1)
    >>> m.logposterior
    -1.8628770664093453
    >>> m.current_stats
    {'logjacobian': 0.0, 'loglikelihood': -1.8628770664093453, 'logprior': 0.0}
```
Here is how you would do the same using the `CallModel` wrapper:
```
    >>> m = TestNormal(['x', 'y'])
    >>> call_model = CallModel(m, 'logposterior')
     >>> call_model([0.1, -0.2])
    (-1.8628770664093453, (0.0, 0.0, -1.8628770664093453))
    >>> call_model.return_all_stats = False
    >>> call_model([0.1, -0.2])
    -1.8628770664093453
```
The `CallModel` wrapper promotes all attributes of the given model instance to its own name space so it can be accessed in the same way as the underlying model.

Some other changes in this PR:
 * I've abstracted `BaseModel` and `BaseDataModel` (which was renamed from `DataModel`). The inheritance is still chained (e.g., `GaussianNoise` inherits from `BaseDataModel` which inherits from `BaseModel`), but if we want more abstraction that can be done in the future.
 * `BaseDataModel` has been moved to its own file, `base_data.py` to make `base.py` smaller.
 * Noise-dependent attributes, like `loglr`, `lognl`, and `logplr` have been moved to `BaseDataModel`, as they really have no meaning without data.
 * `BaseModel` therefore only requires a `_loglikelihood` method to be defined, which all of the analytic models do.
 * `BaseDataModel` adds that a `_loglr` and `_lognl` method also need to be defined.
 * I was unsure of was what to have `MarginalizedPhaseGausssianNoise` calculate for the complex loglr in each detector. So for now, it simply doesn't calculate them.
 * I've updated the documentation for the classes to reflect these changes. In the process I removed a bunch of background information; some more details can be added back in the future.

Commit:
update docs

Commit:
fix bugs related to config loading

Commit:
fix more typos

Commit:
more typos

Commit:
update plotting codes

Commit:
update plot_movie

Commit:
update docs

Commit:
fix typos

Commit:
docs update

Commit:
more docs updates

Commit:
fix whitespace

Commit:
fix unit test

cdcapano commented:
@vivienr I think this is ready to merge now. In addition to the travis tests, I've done quick runs of the command-line examples.

vivienr reviewed (approved):
Looks good, and we can build upon this at the F2F.
